scripts/normalize-by-median.py

```diff
@@ -55,166 +48,111 @@ def WithDiagnostics(ifile, batch_size, fp, paired, norm):
                 report_fp.flush()
 
-        # If in paired mode, check that the reads are properly interleaved
-        if paired:
-            if not check_is_pair(batch[0], batch[1]):
-                raise IOError('Error: Improperly interleaved pairs \
-                    {b0} {b1}'.format(b0=batch[0].name, b1=batch[1].name))
+        yield record
+
+    # per file diagnostic output
+    if norm.total == 0:
+        print('SKIPPED empty file ' + ifilename, file=sys.stderr)
+    else:
+        print('DONE with {inp}; kept {kept} of {total} or {perc:2}%'
+              .format(inp=ifilename, kept=norm.total - norm.discarded,
+                      total=norm.total, perc=int(100. - norm.discarded /
+                                                 float(norm.total) * 100.)),
+              file=sys.stderr)
 
-        yield batch
+    if fp:
+        print("{total} {kept} {discarded}"
+              .format(total=norm.total, kept=norm.total - norm.discarded,
+                      discarded=1. - (norm.discarded / float(norm.total))),
+              file=fp)
+        fp.flush()
 
 
 class Normalizer(object):
-    def __init__(self, desired_coverage, htable, report_fp=None):
+    def __init__(self, desired_coverage, htable, report_fp=None,
+                 force_single=False):
         self.htable = htable
         self.desired_coverage = desired_coverage
         self.report_fp = report_fp
 
         self.total = 0
         self.discarded = 0
-        self.corrupt_files = []
 
-    def __call__(self, input_filename, force_paired=False):
-        seq = ""
+    def __call__(self, reader):
 
         desired_coverage = self.desired_coverage
         ksize = self.htable.ksize()
 
-        batch_size = 1
-        if force_paired:
-            batch_size = 2
-
-        for batch in WithDiagnostics(input_filename, batch_size,
-                                     self.report_fp, force_paired, self):
+        for index, is_paired, read0, read1 in reader:
             passed_filter = False
-            passed_length = True
 
-            for record in batch:
-                if len(record.sequence) < ksize:
-                    passed_length = False
-                    continue
+            self.total += 1
+
+            if is_paired:
+                self.total += 1
 
+            batch = []
+            batch.append(read0)
+            if read1 is not None:
+                batch.append(read1)
+
+            for record in batch:
                 seq = record.sequence.replace('N', 'A')
                 med, _, _ = self.htable.get_median_count(seq)
 
                 if med < desired_coverage:
                     passed_filter = True
 
-            if passed_length and passed_filter:
+            if passed_filter:
                 for record in batch:
                     seq = record.sequence.replace('N', 'A')
                     self.htable.consume(seq)
-                    yield record
+                    yield record, index
```

Comment:
would suggest removing index (which is not incremented here) and using:

    for index, record in enumerate(norm(reader)):

which is a standard Python pattern.